aten/src/ATen/native/mps/operations/Distributions.mm

```diff
@@ -341,6 +342,44 @@ Tensor normal_mps(const Tensor& mean, const Tensor& std, c10::optional<Generator
                                       "exponential_mps_:" + std::to_string(lambda), random_op_block);
 }
 
+Tensor& randperm_out_mps(int64_t n, c10::optional<Generator> generator, Tensor& result) {
+  if (!MPSDevice::getInstance()->macOS_13_0()) {
+    TORCH_WARN_ONCE("MPS: randperm op is supported natively starting from macOS 13.0. ",
+                    "Falling back on CPU. This may have performace implications.");
+
+    result = result.to("cpu");
+    result = at::randperm_out(result, n).to("mps");
+    return result;
+  }
+
+  using namespace mps;
+  TORCH_CHECK(n >= 0, "n must be non-negative, got", n);
+  TORCH_CHECK(!generator.has_value() || (generator.has_value() && result.device() == generator->device()), "Expected a '", result.device(), "' generator device but found '", generator->device(), "'");
+  check_supported_max_int_with_precision(n, result);
+
+  result.resize_({n});
+  if (n == 0) {
+    return result;
+  }
+
+  mps::RandomOpBlock random_op_block = ^RandomOpFn(cachedGraph, randomTensor) {
+    MPSGraph* mpsGraph = cachedGraph->graph();
+    MPSGraphTensor* argsortTensor = [mpsGraph argSortWithTensor:randomTensor
+                                                           axis:0
+                                                           name:nil];
+    if (result.scalar_type() != kInt) {
+      argsortTensor = [mpsGraph castTensor:argsortTensor
+                                    toType:getMPSDataType(result.scalar_type())
```

Comment:
For casting to integer, there's no need to add this cast here. There's already one embedded in random_mps_impl(). You just need to pass int64_t to its template and it will cast output to `MPSDataTypeInt32`, like this: `mps::random_mps_impl<int64_t>()`.

Reply:
I think result could be of any integer type here (e.g int8, int16, int32, int64) - I think it's very rare though when it could be some other dtype than int64. Is it possible to pass directly the scalar type into the template (e.g without hardcoding it in random_mps_impl<`int64_t`>?

Reply:
oh nvm, int64_t actually works fine here. I've updated it

Reply:
The MPS doesn't support int64 for random ops (only Int32 supported), and the int64_t that you pass to template is just a hint to use MPSDataTypeInt32.